### PR TITLE
Introduce @ResourceParameter and MimeTypes for Trainers and Annotators 

### PR DESCRIPTION
**Pull request by schrieveslaach**
The `@ModelTrainerCapability` annotation can be used in order to load trainer classes through reflection.

### CODE REVIEW

**Comment by reckart** [comment]
Thanks for the PR. Looks interesting. There are many whitespace only changes... I'll soon add a checkstyle configuration to DKPro Core too. Meanwhile, would you mind reverting all whitespace-only changes to facilitate seeing the "meat"?

**Comment by schrieveslaach** [comment]
@reckart, all whitespace changes are reverted.

**Comment by reckart** [comment]
Sorry for not getting back earlier. I have been thinking about this and have several comments:

* `@ModelTrainerCapability` assumes that for each trainer component, there is only a single component which could use the created model. That is not always true. E.g. DKPro Core currently has two different implementations of CoreNLP wrappers, one in the `stanfordnlp` module and one in the `corenlp` module. You might consider this a freak case, but I think it would be better to assume that there can be multiple consumers.
* Instead of a direct reference from trainer to annotator component, I think it would be better to introduce a kind of mime-type for models and then mark trainers as producers of this type of files and annotators as consumers of the respective types. The existing uimaFIT `@MimeTypeCapability` could maybe be used for that.
* The `type` should IMHO refer to UIMA type names from the DKPro Core type system. What does it refer to? To the layer(s) used as features/labels training the model or the layer(s) annotated by the model?

**Comment by schrieveslaach** [comment]
@reckart, thank you very much for your suggestions. Those suggestions make totally sense and my client code is much better. Have a look at the new version. 

I'm not sure about the mimetype format. What do you think?

**Comment by schrieveslaach** [comment]
BTW, what are the differences between StanfordNLP and CoreNLP?

**Comment by reckart** [comment]
**Wrt.  StanfordNLP vs. CoreNLP**: The difference between the DKPro Core StanfordNLP and CoreNLP modules is this:
* the `CoreNLP` module uses the API exposed by Stanford CoreNLP
* the `StanfordNLP` module uses APIs that predate Stanford CoreNLP and come from a time where the Stanford NLP tools were still distributed only separately as different software packages. At that time, there was no nice unified API as CoreNLP provides it.

We still have both because there are subtle differences. Eventually, the `StanfordNLP` should go away and is fully replaced by the `CoreNLP` module.

**Wrt. the mime-type format:** We have a list of mime-types in [MimeTypes](https://github.com/dkpro/dkpro-core/blob/master/dkpro-core-api-parameter-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/api/parameter/MimeTypes.java). I think it would be good to add the mime-types for the models there as well.

They should look similar to "normal" mime-types. E.g. `application/x.org.dkpro.core.stanfordnlp.ner`. The pattern I would use is `application/x.org.dkpro.core.<short-module-name>.<model type>` there `model type` is what we also use in the model artifact names, e.g. `parser`, `ner`, `tagger` etc.

**@ModelTrainerCapability:** is it still required? What do you need it for?

**Wrt. new module:** I think the `api.parameter` should be an adequate place for this kind of addition.

**Comment by schrieveslaach** [comment]
Thanks for the clarification on StanfordNLP vs. CoreNLP. 😉 

I refined the mimetype definition according to your suggestions, see 30c1714.

`@ModelTrainerCapability` is still required because I need to figure out which parameter identifies the model location. For annotators which require only one model (e.g. `OpenNlpNamedEntityRecognizer`) I can rely on the convention that this parameter has the name `ComponentParameters.PARAM_MODEL_LOCATION`. However, for the `OpenNlpSegmenter` I cannot make this assumption because it uses two parameters. Or is there any other convention which I'm currently missing?

Let me give you some more details on what I'm doing: During my Ph.D. thesis I wrote a Maven plugin to train and evaluate domain-specific NLP models and this plugin can be used to integrate this process into a CI pipeline. During our project we created our own little NLP pipeline wrapper because we needed some features which DKProCore did not provide (I asked you about some required features on the GSCL 2013 in Darmstadt on a workshop 😉). Now, I'm finishing my thesis and I'm making the stuff available as OpenSource and I want to use DKProCore at the core (now most of the required features are available 👍). Maybe, more of my work could be integrated into DKProCore?

**Comment by reckart** [comment]
**Wrt. @ModelTrainerCapability** - since we have "mime-types" for the different model types now, how about annotating the model parameters with the mimetypes instead? E.g.

```
	/**
	 * Load the segmentation model from this location instead of locating the model automatically.
	 */
	public static final String PARAM_SEGMENTATION_MODEL_LOCATION = ComponentParameters.PARAM_SEGMENTATION_MODEL_LOCATION;
	@ConfigurationParameter(name = PARAM_SEGMENTATION_MODEL_LOCATION, mandatory = false)
	@ResourceParameter(mimeType=APPLICATION_X_OPENNLP_SENT)
	protected String segmentationModelLocation;
```

I think it would make more sense to maintain this information in the consuming component, not in the trainer component. E.g. if you add a new component that could consume a given model type, then you would currently have to update the trainer component. That should not be necessary. The component itself should be able to declare that it consumes a certain model type and by which parameter it does so.

Do you have your work already somewhere on GitHub? What other aspects do you have in mind for integration?

**Comment by schrieveslaach** [comment]
I created this annotation in the `api.parameter` package and deleted the `ModelTrainerCapability`. This scheme works for my use case.

The code is available on the [Gitlab server of our university](https://git.noc.fh-aachen.de/marc.schreiber/Towards-Effective-NLP-Application-Development). I think that the ODT reader and writer could be integrated into dkpro core (our student is still working on this).

**Comment by schrieveslaach** [comment]
@reckart, any thoughts or comments on the current state of the PR?

**Comment by reckart** [comment]
@schrieveslaach I'm looking forward to coming back to it soon, yes. I've been very busy over at WebAnno in the last days as it is wrapping up for a next release.

**Comment by schrieveslaach** [comment]
@reckart, thanks for the update! 😉

**Comment by reckart** [comment]
Jenkins, could you test this please?

**Comment by schrieveslaach** [comment]
@reckart, why does the check fail?

**Comment by reckart** [comment]
```
Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:2.10:analyze-only (default-cli) on project de.tudarmstadt.ukp.dkpro.core.opennlp-asl:

Cannot analyze dependencies: Trying to force use of dependencies which are declared but already detected as used: [de.tudarmstadt.ukp.dkpro.core:de.tudarmstadt.ukp.dkpro.core.api.parameter-asl]
```

A dependency which is presently **enforced** as being used in the POM is now actually **detected** as being used. The build fails until the exception for this dependency is removed.

Cf. https://github.com/dkpro/dkpro-core/blob/master/dkpro-core-opennlp-asl/pom.xml#L455

**Comment by schrieveslaach** [comment]
Nothing I can fix in this PR, correct? 😅

**Comment by reckart** [comment]
Well, you could. You would remove these lines from the POM:

```
            <!--
                - Maven doesn't detect the parameters module to be used, be we actually
                - do require it as a compile dependency and also at runtime, so we
                - cannot set it to scope provided. Need to tell Maven to ignore it here.
              -->
              <usedDependency>de.tudarmstadt.ukp.dkpro.core:de.tudarmstadt.ukp.dkpro.core.api.parameter-asl</usedDependency>
```

**Comment by schrieveslaach** [comment]
Jenkins, could you test this please?

**Comment by reckart** [comment]
Jenkins, test this please.

@schrieveslaach Sorry, but Jenkins currently only listens to me ;)

**Comment by reckart** [comment]
Jenkins, can you test this please.

**Comment by reckart** [comment]
@schrieveslaach The Jenkins config to test PRs is new... for some reason the last build failed to update the commit status. That's why I am trying another Jenkins build again.

File: dkpro-core-opennlp-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/opennlp/OpenNlpChunker.java [hunk 1]
```diff
@@ -55,6 +58,7 @@
 /**
  * Chunk annotator using OpenNLP.
  */
+@MimeTypeCapability(MimeTypes.APPLICATION_X_OPENNLP_CHUNK)
```
Comment (reckart) [inline dkpro-core-opennlp-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/opennlp/OpenNlpChunker.java]: I would reserve the MimeTypeCapability to input files and output files. The ResourceParameter annotation below on the modelLocation parameter should be sufficient.

File: dkpro-core-api-parameter-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/api/parameter/ResourceParameter.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+     * The MIME types supported by this component.
+     *
+     * @return the supported MIME types
+     */
+    String[] value() default NO_DEFAULT_VALUE;
+
+    /**
+     * Provides the default value that tells the ResourceParameter that no value has been provided.
+     */
+    public static final String NO_DEFAULT_VALUE = "org.apache.uima.fit.descriptor.MimeTypeCapability.NO_DEFAULT_VALUE";
```
Comment (reckart) [inline dkpro-core-api-parameter-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/api/parameter/ResourceParameter.java]: That would be `de.tudarmstadt.ukp.dkpro.core.api.parameter.NO_DEFAULT_VALUE`, right? However, I would tend towards using `{}` as the default value in line 33 and not using this uimaFIT-ismic hack.

File: dkpro-core-opennlp-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/opennlp/OpenNlpChunker.java (hunk 1 above)
Comment (schrieveslaach) [inline dkpro-core-opennlp-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/opennlp/OpenNlpChunker.java]: Currently, I rely on the `@MimeTypeCapability` in order to find the link between trainer and annotator. If I remove this annotation from the classes, I cannot construct this relation.

Should I replace this annotation with new one, e.g. `@ResourceParameterProducer(MimeTypes.APPLICATION_X_OPENNLP_CHUNK)`?

File: dkpro-core-api-parameter-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/api/parameter/ResourceParameter.java (hunk 2 above)
Comment (schrieveslaach) [inline dkpro-core-api-parameter-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/api/parameter/ResourceParameter.java]: I will fix it in the next commit.

File: dkpro-core-opennlp-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/opennlp/OpenNlpChunker.java (hunk 1 above)
Comment (reckart) [inline dkpro-core-opennlp-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/opennlp/OpenNlpChunker.java]: I would leave the `@MimeTypeCapability(MimeTypes.APPLICATION_X_OPENNLP_CHUNK)` on the trainer and retain only the `@ResourceParameter(MimeTypes.APPLICATION_X_OPENNLP_CHUNK)` on the annotator. Doesn't that give you the desired relation?

Comment (schrieveslaach) [inline dkpro-core-opennlp-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/opennlp/OpenNlpChunker.java]: This works for me. :v:

**Comment by schrieveslaach** [comment]
@reckart, I improved the code according to your suggestions. ☺️

**Comment by reckart** [comment]
Thx :) 👍

**Comment by schrieveslaach** [comment]
Cool, thanks for merging! ✌️

**Comment by reckart** [comment]
@schrieveslaach The ODT reader/writer could be an interesting contribution. Best we would discuss it in a separate issue.

**Comment by schrieveslaach** [comment]
@reckart, that sounds reasonable. I will create a new issue when our trainee finished the core of it.

**Pull request merged**